prow/pjutil/filter.go

```diff
@@ -80,35 +81,92 @@ func AvailablePresubmits(changes config.ChangedFilesProvider, org, repo, branch
 //  - we know that the presubmit is forced to run
 //  - what the default behavior should be if the presubmit
 //    runs conditionally and does not match trigger conditions
-type Filter func(p config.Presubmit) (shouldRun bool, forcedToRun bool, defaultBehavior bool)
+type Filter interface {
+	ShouldRun(p config.Presubmit) (shouldRun bool, forcedToRun bool, defaultBehavior bool)
+	Name() string
+}
+
+type ArbitraryFilter struct {
+	override func(p config.Presubmit) (shouldRun bool, forcedToRun bool, defaultBehavior bool)
+	name     string
+}
+
+func NewArbitraryFilter(override func(p config.Presubmit) (shouldRun bool, forcedToRun bool, defaultBehavior bool), name string) *ArbitraryFilter {
+	return &ArbitraryFilter{override: override, name: name}
+}
+
+func (af *ArbitraryFilter) ShouldRun(p config.Presubmit) (shouldRun bool, forcedToRun bool, defaultBehavior bool) {
+	return af.override(p)
+}
+
+func (af *ArbitraryFilter) Name() string {
+	return af.name
+}
 
 // CommandFilter builds a filter for `/test foo`
-func CommandFilter(body string) Filter {
-	return func(p config.Presubmit) (bool, bool, bool) {
-		return p.TriggerMatches(body), p.TriggerMatches(body), true
+type CommandFilter struct {
```

Comment:
nit: We don't need separate structs to implement these, `ArbitraryFilter` could be used for all of them. Fields like body could just be captured variables in the `override` function, captured from the parameters of the `New*Filter` functions.

Reply:
I would consider ArbitraryFilter a "lazy" struct for convenience purpose, for example used in unit test. Make every Filter to use ArbitraryFilter technically could work, but imo it would complicate the implementation by considering different cases, for example custom Name function for CommandFilter. I've added comment explaining why ArbitraryFilter exists